### PR TITLE
Add omp_for_schedule and omp_for_nowait_schedule policies

### PR DESCRIPTION
**Pull request by willkill07**
# Summary

- This PR is a feature
- It does the following (modify list as needed):
  - Modifies/refactors OpenMP policies for CPU by augmenting schedule types
  - Adds Guided, Runtime, Auto, and Dynamic scheduling options per the discussion at the last RAJA dev meeting

# Design review

This introduces two new end-user policies: `omp_for_schedule_exec<T>` and `omp_for_nowait_schedule_exec<T>` where:
 -  `T` is a scheduling policy
 - `omp_for_schedule_exec` only specifies the scheduling policy
 - `omp_for_nowait_schedule_exec` also includes the nowait clause on the loop

A scheduling policy is one of:
- `RAJA::policy::omp::Dynamic<N>`
- `RAJA::policy::omp::Static<N>`
- `RAJA::policy::omp::Guided<N>`
- `RAJA::policy::omp::Auto`
- `RAJA::policy::omp::Runtime`

This uses tag dispatch in an internal namespace to `RAJA::policy::omp` to reduce lookup.

Any feedback is welcome.

# TODO
- [x] Handle compiler workarounds for MSVC
- [x] Write unit/functional tests for each policy
- [x] Add RAJA::policy::omp schedule clauses to documentation

### CODE REVIEW

**Comment by willkill07** [comment]
@rhornung67 and @trws -- any feedback here would be good. I'm flexible in terms of any reasonable solution. I know this won't scale with a product of options.

**Comment by trws** [comment]
I need to think about this a bit, but am fine with the concept.  First blush I’m thinking we can collapse this a little bit so we don’t have quite such a profusion without losing much if we can figure out how to spell it.  Reasoning:

* auto is what you get when you don’t use a schedule clause
* static is unique and needs to stand alone, could have dynamic or static chunk size
* dynamic, guided and runtime tend to all be implemented by the same fully dynamic code path, so we could cover all three of these and static with a dynamic chunk size with a single impl that uses `schedule(runtime, chunk)` with `omp_set_schedule`

The no wait is actually where I’m wondering what we should do.  I keep thinking we’d do well to have some expected semantics from Forall, but we really kinda don’t, it’s just defined as “whatever the policy does in the underlying model” and that may become hard to reason about as the async kernels stuff starts to get in heavier use.

For what it’s worth, I’d be fine with taking this as is if it’s a use-case we want to support.  It just has me thinking about how much of the OpenMP interface we want to surface through raja to the user, especially where it’s only really usable inside a region.

**Comment by willkill07** [comment]
@trws -- thanks

I agree with auto and the importance of static. I'm also viewing this as a testbed for any directive-based programming model where we have various directives and clauses.

The single code path for dynamic / guided with `omp_set_schedule` may incur overhead in target offload kernels. I'm not sure what the best approach is.

NVIDIA's new nvc++ compiler looks to be a promising candidate for modern C++ and directive-based abstractions with OpenACC.

File: include/RAJA/policy/openmp/policy.hpp [hunk 1]
```diff
@@ -76,12 +118,7 @@ struct omp_for_nowait_exec
 };
 
 template <unsigned int N>
-struct omp_for_static : make_policy_pattern_launch_platform_t<Policy::openmp,
-                                                              Pattern::forall,
-                                                              Launch::undefined,
-                                                              Platform::host,
-                                                              omp::For,
-                                                              omp::Static<N>> {
+struct omp_for_static : omp_for_schedule<omp::Static<N>> {
```
Comment (MrBurmark) [inline include/RAJA/policy/openmp/policy.hpp:121]: Could this be an alias to avoid generating another class?

Comment (rhornung67) [reply include/RAJA/policy/openmp/policy.hpp:121]: You mean the omp::Static part?  Will was working on a PR to clean up some of this. We briefly discussed at the RAJA meeting earlier this week. I'll defer to the OpenMP experts.

File: include/RAJA/policy/openmp/forall.hpp [hunk 2]
```diff
@@ -68,52 +67,150 @@ RAJA_INLINE void forall_impl(const omp_parallel_exec<InnerPolicy>&,
...
 {
-  RAJA_EXTRACT_BED_IT(iter);
-#pragma omp for nowait
-  for (decltype(distance_it) i = 0; i < distance_it; ++i) {
-    loop_body(begin_it[i]);
+
+  /// Tag dispatch for omp forall
+
+  template <typename Iterable, typename Func, unsigned int ChunkSize>
+  RAJA_INLINE void forall_impl(const ::RAJA::policy::omp::Auto&,
```
Comment (willkill07) [inline include/RAJA/policy/openmp/forall.hpp:81]: auto is our "old" default

File: include/RAJA/policy/openmp/forall.hpp [hunk 3]
```diff
@@ -68,52 +67,150 @@ RAJA_INLINE void forall_impl(const omp_parallel_exec<InnerPolicy>&,
...
+      loop_body(begin_it[i]);
+    }
   }
-}
 
-///
-/// OpenMP parallel for policy implementation
-///
+  template <typename Iterable, typename Func, unsigned int ChunkSize>
+  RAJA_INLINE void forall_impl(const ::RAJA::policy::omp::Static<ChunkSize>&,
```
Comment (willkill07) [inline include/RAJA/policy/openmp/forall.hpp:93]: explicit specialization for Static<ChunkSize>

File: include/RAJA/policy/openmp/forall.hpp [hunk 4]
```diff
@@ -68,52 +67,150 @@ RAJA_INLINE void forall_impl(const omp_parallel_exec<InnerPolicy>&,
...
+    #pragma omp for schedule(static, ChunkSize)
+    for (decltype(distance_it) i = 0; i < distance_it; ++i) {
+      loop_body(begin_it[i]);
+    }
+  }
 
-template <typename Iterable, typename Func>
-RAJA_INLINE void forall_impl(const omp_for_exec&,
+  template <typename Iterable, typename Func>
+  RAJA_INLINE void forall_impl(const ::RAJA::policy::omp::Runtime&,
```
Comment (willkill07) [inline include/RAJA/policy/openmp/forall.hpp:105]: explicit specialization for Runtime

File: include/RAJA/policy/openmp/forall.hpp [hunk 5]
```diff
@@ -68,52 +67,150 @@ RAJA_INLINE void forall_impl(const omp_parallel_exec<InnerPolicy>&,
...
+  {
+    RAJA_EXTRACT_BED_IT(iter);
+    #pragma omp for schedule(runtime)
+    for (decltype(distance_it) i = 0; i < distance_it; ++i) {
+      loop_body(begin_it[i]);
+    }
+  }
+
+  // dynamic & guided
+  template <typename Policy, typename Iterable, typename Func, unsigned int ChunkSize>
```
Comment (willkill07) [inline include/RAJA/policy/openmp/forall.hpp:117]: template argument deduction for runtime and guided -- using `omp_set_schedule` to choose what to do

File: include/RAJA/policy/openmp/forall.hpp [hunk 6]
```diff
@@ -68,52 +67,150 @@ RAJA_INLINE void forall_impl(const omp_parallel_exec<InnerPolicy>&,
...
+RAJA_INLINE void forall_impl(const omp_for_schedule<Schedule>&,
                              Iterable&& iter,
                              Func&& loop_body)
 {
-  RAJA_EXTRACT_BED_IT(iter);
-#pragma omp for
-  for (decltype(distance_it) i = 0; i < distance_it; ++i) {
-    loop_body(begin_it[i]);
-  }
+  internal::forall_impl(Schedule{}, std::forward<Iterable>(iter), std::forward<Func>(loop_body));
```
Comment (willkill07) [inline include/RAJA/policy/openmp/forall.hpp:205]: internal tagged dispatch to schedule specified

File: include/RAJA/policy/openmp/policy.hpp [hunk 7]
```diff
@@ -22,14 +22,32 @@
...
 {
-
 namespace omp
 {
 
+namespace internal
+{
+    template <omp_sched_t Sched, int Chunk>
+    struct Schedule {
+        constexpr static omp_sched_t schedule = Sched;
```
Comment (willkill07) [inline include/RAJA/policy/openmp/policy.hpp:38]: `schedule` will be negative if `Auto`
`chunk_size` will be negative if "default"

File: include/RAJA/policy/openmp/policy.hpp [hunk 8]
```diff
@@ -42,10 +60,23 @@ struct For {
 struct NoWait {
 };
 
-template <unsigned int ChunkSize>
-struct Static : std::integral_constant<unsigned int, ChunkSize> {
+static constexpr int default_chunk_size = -1;
```
Comment (willkill07) [inline include/RAJA/policy/openmp/policy.hpp:63]: This is a *public* interface option for users. The OpenMP runtime says that a "default" will be used when a passed value into `omp_set_schedule` is negative

File: include/RAJA/policy/openmp/policy.hpp [hunk 9]
```diff
@@ -42,10 +60,23 @@ struct For {
...
+
+struct Auto : private internal::Schedule<omp_sched_auto, default_chunk_size>{
```
Comment (willkill07) [inline include/RAJA/policy/openmp/policy.hpp:65]: `Auto`, `Static`, and `Runtime` are new types so tagged dispatch works correctly

File: include/RAJA/policy/openmp/policy.hpp [hunk 10]
```diff
@@ -55,52 +86,58 @@ struct Static : std::integral_constant<unsigned int, ChunkSize> {
 //////////////////////////////////////////////////////////////////////
 //
 
+namespace exec {
```
Comment (willkill07) [inline include/RAJA/policy/openmp/policy.hpp:89]: I placed all execution policy names in the exec namespace so users can autocomplete `RAJA::policy::omp::` to get the various schedules and `default_chunk_size`

File: include/RAJA/policy/openmp/policy.hpp [hunk 11]
```diff
@@ -55,52 +86,58 @@ struct Static : std::integral_constant<unsigned int, ChunkSize> {
...
-    : make_policy_pattern_t<Policy::openmp, Pattern::forall, omp::For> {
+template <typename Schedule>
+struct omp_for_nowait_schedule : make_policy_pattern_launch_platform_t<Policy::openmp,
+                                                              Pattern::forall,
+                                                              Launch::undefined,
+                                                              Platform::host,
+                                                              omp::For,
+                                                              omp::NoWait,
+                                                              Schedule> {
+    static_assert(std::is_base_of<::RAJA::policy::omp::internal::Schedule, Schedule>::value,
```
Comment (willkill07) [inline include/RAJA/policy/openmp/policy.hpp:106]: Compile-time error for users who don't specify a correct schedule

File: include/RAJA/policy/openmp/policy.hpp [hunk 12]
```diff
@@ -55,52 +86,58 @@ struct Static : std::integral_constant<unsigned int, ChunkSize> {
...
                                                               Launch::undefined,
                                                               Platform::host,
                                                               omp::For,
-                                                              omp::Static<N>> {
+                                                              Schedule> {
+    static_assert(std::is_base_of<::RAJA::policy::omp::internal::Schedule, Schedule>::value,
+        "Schedule must be one of: Auto|Runtime|Static|Dynamic|Guided");
 };
 
+using omp_for_exec = omp_for_schedule<Auto>;
```
Comment (willkill07) [inline include/RAJA/policy/openmp/policy.hpp:122]: All other policies now alias to `omp_for_schedule` or `omp_for_nowait_schedule`

File: include/RAJA/policy/openmp/forall.hpp [hunk 13]
```diff
@@ -68,52 +67,150 @@ RAJA_INLINE void forall_impl(const omp_parallel_exec<InnerPolicy>&,
...
+
+  // dynamic & guided
+  template <typename Policy, typename Iterable, typename Func, unsigned int ChunkSize>
+  RAJA_INLINE void forall_impl(const Policy&,
+                               Iterable&& iter,
+                               Func&& loop_body)
+  {
+    omp_sched_t prev_sched;
+    int prev_chunk;
+    omp_get_schedule(&prev_sched, &prev_chunk);
```
Comment (willkill07) [inline include/RAJA/policy/openmp/forall.hpp:124]: @trws is preserving the old schedule settings important? I have this added in and don't think it should add too much to the implementation (only used for `Guided` and `Dynamic`)

**Comment by willkill07** [comment]
The HIP CI tests reveal something not-so-great:

```
In file included from /home/raja/include/RAJA/pattern/kernel/Region.hpp:23:
938/home/raja/include/RAJA/policy/openmp/policy.hpp:25:10: fatal error: 'omp.h' file not found
939#include <omp.h>
```

Is there a reason why `Region.hpp` is even including `openmp/policy.hpp` when nothing explicitly depends on it?

**Comment by willkill07** [comment]
Additionally, MSVC only supports OpenMP 2.0 -- the `omp.h` functions and types:
* `omp_set_schedule`
* `omp_get_schedule`
* `omp_sched_t`

Only were added with OpenMP 3.0. Should I add an `#if _OPENMP > 200805` guard and limit the code path for what's visible?

**Comment by rhornung67** [comment]
> Additionally, MSVC only supports OpenMP 2.0 -- the `omp.h` functions and types:
> 
>     * `omp_set_schedule`
> 
>     * `omp_get_schedule`
> 
>     * `omp_sched_t`
> 
> 
> Only were added with OpenMP 3.0. Should I add an `#if _OPENMP > 200805` guard and limit the code path for what's visible?

I think that adding the guard makes sense.  However, it seems odd that the compiler does not support OpenMP 3.0

**Comment by rhornung67** [comment]
> The HIP CI tests reveal something not-so-great:
> 
> ```
> In file included from /home/raja/include/RAJA/pattern/kernel/Region.hpp:23:
> 938/home/raja/include/RAJA/policy/openmp/policy.hpp:25:10: fatal error: 'omp.h' file not found
> 939#include <omp.h>
> ```
> 
> Is there a reason why `Region.hpp` is even including `openmp/policy.hpp` when nothing explicitly depends on it?

Probably an oversight after some code regorg. If that header is not needed, please remove its inclusion. Hopefully, it won't result in a cascade of other files relying on it being included here.

**Comment by willkill07** [comment]
@rhornung67 I made those changes and have a clean build across.

**Documentation**: Added, but a build wasn't triggered on readthedocs. I manually inspected a local build, though
**Testing**: Added to forall and atomic testing, but it's only enabled with `RAJA_TEST_EXHAUSTIVE`

@trws I think changes are at a point for a review. Not sure who else would be interested in reviewing this.

File: include/RAJA/policy/openmp/forall.hpp (hunk 13 above)
Comment (trws) [reply include/RAJA/policy/openmp/forall.hpp]: As long as it only exists for guided and dynamic I have no issues with it at all.  The dynamic code paths tend to have a good deal more overhead for startup anyway, and are less frequently used.  There may be an implementation at some point where we decide to care, but none exists today that I know of.

**Comment by willkill07** [comment]
> I'm pretty happy. @willkill07, do you have time to update to match main?

@trws Wouldn't this go into `develop`? -- or should change the target merge to `main`?

**Comment by rhornung67** [comment]
@willkill07 please merge into develop as your PR is set up to do.

**Comment by willkill07** [comment]
@rhornung67 this can be merged once checks pass (I cannot merge)

**Comment by rhornung67** [comment]
Will do.

**Comment by rhornung67** [comment]
@willkill07 merging your OpenMP policy PR.

**Pull request merged**